core/src/identity/ecdsa.rs

```diff
@@ -0,0 +1,161 @@
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! ECDSA keys with secp256r1 curve support.
+
+use super::error::DecodingError;
+use core::fmt;
+use p256::ecdsa::{
+    signature::{Signer, Verifier},
+    Signature, SigningKey, VerifyingKey,
+};
+
+/// An ECDSA keypair.
+#[derive(Clone)]
+pub struct Keypair {
+    secret: SecretKey,
+    public: PublicKey,
+}
+
+impl Keypair {
+    /// Generate a new random ECDSA keypair.
+    pub fn generate() -> Keypair {
+        Keypair::from(SecretKey::generate())
+    }
+
+    /// Sign a message using the private key of this keypair.
+    pub fn sign(&self, msg: &[u8]) -> Vec<u8> {
+        self.secret.sign(msg)
+    }
+
+    /// Get the public key of this keypair.
+    pub fn public(&self) -> &PublicKey {
+        &self.public
+    }
+
+    /// Get the secret key of this keypair.
+    pub fn secret(&self) -> &SecretKey {
+        &self.secret
+    }
+}
+
+impl fmt::Debug for Keypair {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Keypair")
+            .field("public", &self.public())
+            .finish()
+    }
+}
+
+/// Promote an ECDSA secret key into a keypair.
+impl From<SecretKey> for Keypair {
+    fn from(secret: SecretKey) -> Keypair {
+        let public = PublicKey(VerifyingKey::from(&secret.0));
+        Keypair { secret, public }
+    }
+}
+
+/// Demote an ECDSA keypair to a secret key.
+impl From<Keypair> for SecretKey {
+    fn from(kp: Keypair) -> SecretKey {
+        kp.secret
+    }
+}
+
+/// An ECDSA secret key.
+#[derive(Clone)]
+pub struct SecretKey(SigningKey);
+
+impl SecretKey {
+    /// Generate a new random ECDSA secret key.
+    pub fn generate() -> SecretKey {
+        SecretKey(SigningKey::random(rand::thread_rng()))
+    }
+
+    /// Sign a message with this secret key, producing a DER-encoded ECDSA signature.
+    pub fn sign(&self, msg: &[u8]) -> Vec<u8> {
+        self.0.sign(msg).to_der().as_bytes().to_owned()
+    }
+}
+
+impl fmt::Debug for SecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "SecretKey")
```

Comment:
Over at rust-bitcoin, we settled on including a fingerprint (first couple of bytes of tagged hash) in the debug output to make it possible to differentiate different secret keys. Thoughts on doing the same thing here?

Reply:
I've followed the design of the other key types. I can add the attribute, but then I suggest to do the same for the others.

Can you elaborate for what I should use as private key fingerprint?

Reply:
> I've followed the design of the other key types. I can add the attribute, but then I suggest to do the same for the others.

Leave it then, would probably need more discussion overall :)

> Can you elaborate on what should be used as private key fingerprint?

I'd suggest using the first couple of bytes of a tagged hash (ie `sha256(b"ECDSA-secret-key" + bytes)[0..8]`) but if none of the other secret keys do this, we don't need to add it here.